### PR TITLE
Single package.xml file for both ROS1 and ROS2

### PR DESCRIPTION
**Pull request by Timple**
Single package.xml file for both ROS1 and ROS2

This saves the trouble of moving around files.

Basically this is a combination of the `package_ros1.xml` and the `package_ros2.xml` since they are largely identical.
The deviating parts have a conditional guard as supported by `<package format="3">` since ros kinetic.

### CODE REVIEW

**Comment by Timple** [comment]
Any update on this? It's blocking https://github.com/RoboSense-LiDAR/rslidar_sdk/issues/33

**Comment by Timple** [comment]
If you need me to re-target this to the `dev` branch, please let me know.

**Comment by baoxianzhang** [comment]
ROS1 and ROS2 xml are different in build_depend and build_tool_depend. 
image

**Comment by Timple** [comment]
Yes they are. That is why I put conditional guards around them in this PR

**Comment by baoxianzhang** [comment]
I got it. Good job. Test on ROS1 is OK. I will continue to test on ROS2.

**Comment by Timple** [comment]
Great, let me know the outcome!

**Comment by Timple** [comment]
Subtle ping, I felt like this was getting a bit of traction :slightly_smiling_face:

**Comment by baoxianzhang** [comment]
Test on ROS2 is OK. We will update soon.

**Comment by ronzheng1006** [comment]
Hi Timple, 

It is a cool solution. 

However, some of other clients wish not to mixture ros1's package.xml and ros2's, so we have to keep two separate config files. 

Thanks.

**Comment by Timple** [comment]
That is a bummer. There are guards in the [CMakeLists.txt](https://github.com/RoboSense-LiDAR/rslidar_sdk/blob/654d0e75e353d615a0ca62f1a241a7bced15e296/CMakeLists.txt#L30), guards in the [code](https://github.com/RoboSense-LiDAR/rslidar_sdk/blob/654d0e75e353d615a0ca62f1a241a7bced15e296/node/rslidar_sdk_node.cpp#L69) already. So interesting that a guard in the `package.xml` is too much.

But good that you took a look at the existing customer base before merging!

**Comment by ronzheng1006** [comment]
Hi Timple,

Some clients are still working on Ubuntu 16.04/ROS Kinetic, which doesn't know `depend` unfortunately.
We'll merge this PR after saying bye to Kinetic.

**Comment by Timple** [comment]
It should support `depend` actually. It was introduced for Kinetic here: https://github.com/ros/catkin/pull/1025

But nonetheless, kinetic is EOL: https://dlu.github.io/ros_clock/index.html
Do you have a time-estimate on when you plan to say bye to Kinetic? Because until that time we (and lots of other developers I assume?) need to patch this repository on every deployment.

One could easily tag the current master branch with a `kinetic` tag. All clients using legacy software can keep using that tag indefinitely.

**Comment by Timple** [comment]
I rebased the feature against `dev`. Seems all development efforts are going there.

**Comment by JWhitleyWork** [comment]
ROS Kinetic was EoL in April of 2021. Can we please get this merged?

**Comment by Timple** [comment]
These lidars are decent hardware. However software was unable to run at 10hz in ROS2, can happen.

However: no response to build fixes, no response to PR's improving code performance in ROS2, no response to this PR which improves ROS workflow.
We are evaluating other lidar manufacturers. 

I will keep this PR open for those still interested in RS lidars.

**Comment by JWhitleyWork** [comment]
@Timple I am beginning to agree with your assessment. It's unfortunate because some of thier hardware is pretty decent.

**Comment by rafa-martin** [comment]
Is there any update for this PR?. I would be nice use original repository instead of fork

**Comment by jjannik** [comment]
Is there any update?

**Comment by alexg-k** [comment]
ROS1 for the default package configuration becomes more and more outdated. Please merge the efforts of the community into your products. We have just begun buying your products for evaluation purposes and software support is one part.

**Review by rafa-martin — APPROVED** [review]
Tested in noetic

**Comment by Timple** [comment]
So we actually did find an alternative for the rslidar lidars.

However I'm reviving an old robot and discovered this PR is still not merged...

I retargeted the PR to `main` as now this seems to be where development is going on.

**Comment by rafa-martin** [comment]
@Timple it seems that this branch coudn't be merged

@FelixHuang18 could you check it??

**Comment by Timple** [comment]
@rafa-martin do you have any intention of merging if the conflicts are resolved?

**Comment by rafa-martin** [comment]
Unfortunately, I'm not the owner of this repository. :(

**Comment by Timple** [comment]
Then I'm going to leave it as is...
It's functional as far as I know.
 
It's been three years without a response...

**Comment by FelixHuang18** [comment]
@Timple 
Hi Timple,

Thank you so much for your pull request and your contribution to the project! I want to sincerely apologize for the delay in addressing it.

Unfortunately, due to some recent changes, there are now conflicts that prevent me from merging it directly. Could I kindly ask you to resolve the conflicts and push the updated changes?

Once the conflicts are resolved, I will make it a priority to review, verify, and merge your PR as soon as possible.

I appreciate your understanding and patience, and again, thank you for your contribution!

Best regards,
Felix Huang

**Comment by Timple** [comment]
Here you go!

I do not speak, nor write Chinese. So any help there is appreciated.

File: node/rslidar_sdk_node.cpp
Comment (Timple) [inline node/rslidar_sdk_node.cpp:91]: If you can provide an example on how to provide this via ROS2 parameters, this snippet can probably be gone.

But this does not work at the moment:

```XML
<?xml version="1.0" encoding="UTF-8"?>
<launch>
    <node pkg="rslidar_sdk" exec="rslidar_sdk_node" output="screen" >
        <param name="config_path" value="/path/to/rslidar.yaml" />
    </node>
</launch>
```

Comment (FelixHuang18) [reply node/rslidar_sdk_node.cpp:91]: Is the function you want to implement is to load the configuration path from ROS2 parameters?

Comment (Timple) [reply node/rslidar_sdk_node.cpp:91]: Yes
But the `rslidar_sdk` seems to consist of multiple nodes. So I am unsure how to pass a parameter to this.

Comment (FelixHuang18) [reply node/rslidar_sdk_node.cpp:91]: I modified the start.py file to pass the config_path parameter.
`from launch import LaunchDescription
from launch_ros.actions import Node
from ament_index_python.packages import get_package_share_directory
def generate_launch_description():

    rviz_config=get_package_share_directory('rslidar_sdk')+'/rviz/rviz2.rviz'

    config_file = '/home/felix/felix/ws/github/test/src/rslidar_sdk/config/config_ros2.yaml'

    return LaunchDescription([
        Node(namespace='rslidar_sdk', package='rslidar_sdk', executable='rslidar_sdk_node', output='screen',
        parameters=[{'config_path': config_file}]),
        Node(namespace='rviz2', package='rviz2', executable='rviz2', arguments=['-d',rviz_config])
    ])`
test_load_param

**Comment by Timple** [comment]
Ah, that does work. Not sure where my confusion came from.

There are mutiple nodes however: 
```
$ ros2 node list
/param_handle
/rslidar_points_destination_0
```

But that's out of scope for this PR I guess.

**Comment by Timple** [comment]
Would you like to drop the manual argument parsing? Or would it still be convenient to do this?

**Comment by FelixHuang18** [comment]
Yes, I don't think this code is needed.

**Comment by rafa-martin** [comment]
😄

File: CMakeLists.txt
Comment (FelixHuang18) [inline CMakeLists.txt:66]: In addition, why is the restriction of ros version removed here?

Comment (rafa-martin) [reply CMakeLists.txt:66]: Im not sure, but as this software is C++17 compatible, @Timple set it c++17 as default.

For new ROS2 distros you would need to add it here (iron, jazzy and rolling) if new distros are released.

Also, as you are using cmake version > 3.1. Is better define these variables:
set (CMAKE_CXX_STANDARD 17)
https://cmake.org/cmake/help/latest/variable/CMAKE_CXX_STANDARD.html

Comment (Timple) [reply CMakeLists.txt:66]: > In addition, why is the restriction of ros version removed here?

It's not removed, it is still implicitly there. But the statement overlooked `iron`, `jazzy` and `rolling`. Currently `humble` is the oldest supported distro at the moment: [ros_clock](https://dlu.github.io/ros_clock/index.html). So all can use `c++17`

**Pull request merged**
